core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -63,80 +66,136 @@ public class SharePartitionManager {
+                topicPartitionData, future);
+        addToFetchQueue(shareFetchPartitionData);
+
+        fetchResults();
+        return future;
+    }
+
+    public void addToFetchQueue(ShareFetchPartitionData shareFetchPartitionData) {
+        lock.lock();
+        try {
+            fetchQueue.add(shareFetchPartitionData);
+        } finally {
+            lock.unlock();
+        }
+    }
 
-            replicaManager.fetchMessages(
-                fetchParams,
-                CollectionConverters.asScala(
-                    topicPartitionData.entrySet().stream().map(entry -> new Tuple2<>(entry.getKey(), entry.getValue())).collect(
-                        Collectors.toList())
-                ),
-                QuotaFactory.UnboundedQuota$.MODULE$,
-                responsePartitionData -> {
-                    List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
-                        responsePartitionData);
-                    Map<TopicIdPartition, ShareFetchResponseData.PartitionData> result = new HashMap<>();
-                    responseData.forEach(data -> {
-                        TopicIdPartition topicIdPartition = data._1;
-                        FetchPartitionData fetchPartitionData = data._2;
-
-                        partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition))
-                            .acquire(memberId, fetchPartitionData).whenComplete((acquiredRecords, throwable) -> {
-                                ShareFetchResponseData.PartitionData partitionData = new ShareFetchResponseData.PartitionData()
-                                    .setPartitionIndex(topicIdPartition.partition());
-
-                                if (throwable != null) {
-                                    partitionData.setErrorCode(Errors.forException(throwable).code());
-                                } else {
-                                    // Maybe check if no records are acquired and we want to retry replica
-                                    // manager fetch. Depends on the share partition manager implementation,
-                                    // if we want parallel requests for the same share partition or not.
-                                    partitionData
-                                        .setPartitionIndex(topicIdPartition.partition())
-                                        .setRecords(fetchPartitionData.records)
-                                        .setErrorCode(fetchPartitionData.error.code())
-                                        .setAcquiredRecords(acquiredRecords)
-                                        .setAcknowledgeErrorCode(Errors.NONE.code());
-                                }
-                                result.put(topicIdPartition, partitionData);
-                            });
+    public boolean isFetchQueueEmpty() {
+        lock.lock();
```

Comment:
Why to lock when the queue is thread safe?

Reply:
I was of a similar opinion, but when I checked some other parts of code , we have locks on isEmpty operation as well. Example code snippet [here](https://github.com/confluentinc/kafka/blob/kip-932/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchBuffer.java#L66)